File: go/257.go
package main

import "strconv"

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func binaryTreePaths(root *TreeNode) []string {
  ret := make([]string, 0)
  stn, stp := make([]*TreeNode, 0), make([]string, 0)
  stn = append(stn, root)
  stp = append(stp, strconv.Itoa(root.Val))
  for len(stn) > 0 {
    cur := stn[len(stn)-1]
    stn = stn[:len(stn)-1]
    path := stp[len(stp)-1]
    stp = stp[:len(stp)-1]
    if cur.Left == nil && cur.Right == nil {
      ret = append(ret, path)
    }
    if  cur.Left != nil {
      stn = append(stn, cur.Left)
      stp = append(stp, path + "->" + strconv.Itoa(cur.Left.Val))
    }
    if cur.Right != nil {
      stn = append(stn, cur.Right)
      stp = append(stp, path + "->" + strconv.Itoa(cur.Right.Val))
    }
  }
  return ret
}

File: cpp/257.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  vector<string> binaryTreePaths(TreeNode *root) {
    vector<string> ret;                  // return vector
    vector<TreeNode *> stn;              // stack for nodes
    vector<string> stp;                  // stack for paths
    stn.push_back(root);                 // init
    stp.push_back(to_string(root->val)); // init
    TreeNode *cur;
    string path;
    while (!stn.empty()) {
      cur = stn.back(); // get current node
      stn.pop_back();
      path = stp.back(); // get current path
      stp.pop_back();
      if (cur->left == nullptr && cur->right == nullptr) { // leaf
        ret.push_back(path);
      }
      if (cur->left != nullptr) { // there is left child
        stn.push_back(cur->left);
        stp.push_back(path + "->" + to_string(cur->left->val));
      }
      if (cur->right != nullptr) { // there is right child
        stn.push_back(cur->right);
        stp.push_back(path + "->" + to_string(cur->right->val));
      }
    }
    return ret;
  }
};
